Apps/FlightSoftware/Watchdog2/src/uart.c

```diff
@@ -0,0 +1,393 @@
+
+    /* set P2SEL0.1, and P2SEL0.0 to 0 */
+    P2SEL0 &= ~(BIT0 | BIT1);
+    /* set P2SEL1.1, and P2SEL1.0 to 1 */
+    P2SEL1 |= (BIT0 | BIT1);
+
+    UCA0CTLW0 |= UCSSEL__SMCLK;             // CLK = SMCLK
+    // Baud Rate calculation
+    // 8000000/(16*9600) = 52.083
+    // Fractional portion = 0.083
+    // User's Guide Table 21-4: UCBRSx = 0x04
+    // UCBRFx = int ( (52.083-52)*16) = 1
+    UCA0BRW = 52;                           // 8000000/16/9600
+    UCA0MCTLW |= UCOS16 | UCBRF_1 | 0x4900; // ???
+    UCA0CTLW0 &= ~UCSWRST;                  // Release eUSCI_A0 reset
+    UCA0IE |= UCRXIE;                       // Enable USCI_A0 RX interrupt
+}
+
+/**
+ * disable uart 1
+ */
+void uart1_disable() {
+    UCA1CTLW0 = UCSWRST;                    // Put eUSCI_A1 in reset
+
+    /* set P2SEL0.5, P2SEL0.6 to 0 */
+    P2SEL0 &= ~(BIT5 | BIT6);
+    /* set P2SEL1.5, P2SEL1.6 to 0 */
+    P2SEL0 &= ~(BIT5 | BIT6);
+
+    UCA1IE = 0;
+    UCA1CTLW0 = 0; // clear setup
+}
+
+/**
+ * Initialize UART1 (Lander <-> watchdog)
+ */
+void uart1_init() {
+    uart1tx.idx = 0;
+    uart1tx.used = 0;
+    uart1tx.locked = 0;
+    uart1rx.idx = 0;
+    uart1rx.used = 0;
+
+    UCA1CTLW0 = UCSWRST;                    // Put eUSCI_A1 in reset
+
+    /* Setup for eUSCI_A1 */
+    /* On the MSP430FR5994, pin P2.5 is used for TX and pin P2.6 is used for RX
+     * (ref: pg 95 of datasheet) */
+    /* P2.5 TX: x = 5; P2SEL1.x = 1, P2SEL0.x = 0 */
+    /* P2.6 RX: x = 6; P2SEL1.x = 1, P2SEL0.x = 0 */
+
+    /* set P2SEL0.5, P2SEL0.6 to 0 */
+    P2SEL0 &= ~(BIT5 | BIT6);
+    /* set P2SEL1.5, P2SEL1.6 to 1 */
+    P2SEL1 |= (BIT5 | BIT6);
+
+    UCA1CTLW0 |= UCSSEL__SMCLK;             // CLK = SMCLK
+    // Baud Rate calculation
+    // 8000000/(16*9600) = 52.083
+    // Fractional portion = 0.083
+    // User's Guide Table 21-4: UCBRSx = 0x04
+    // UCBRFx = int ( (52.083-52)*16) = 1
+    UCA1BRW = 52;                           // 8000000/16/9600
+    UCA1MCTLW |= UCOS16 | UCBRF_1 | 0x4900; // ???
+    UCA1CTLW0 &= ~UCSWRST;                  // Release eUSCI_A1 reset
+    UCA1IE |= UCRXIE;                       // Enable USCI_A1 RX interrupt
+}
+
+
+void uart0_tx_nonblocking(uint16_t length, unsigned char *buffer) {
+    uint16_t i=0;
+    unsigned char b;
+    uint16_t curr_idx = uart0tx.idx + uart0tx.used;
+
+    for (i = 0; i < length; i++) {
+        b = buffer[i];
+        uart0tx.buf[curr_idx++] = b;
+        if (curr_idx >= BUFFER_SIZE) curr_idx = 0;
+        uart0tx.used++;
+    }
+
+    /* start interrupts for sending async */
+    UCA0IE |= UCTXIE;
+}
+
+void uart1_tx_nonblocking(uint16_t length, unsigned char *buffer, uint8_t opts) {
+    uint16_t i=0;
+    unsigned char b;
+    uint16_t curr_idx;
+
+    // wait for the buffer to be unlocked
+    while (uart1tx.locked) __delay_cycles(100);
+
+    // disable interrupts to prevent race conditions
+    UCA1IE &= ~UCTXIE;
+    // disable interrupts
+    __bic_SR_register(GIE);
```

Comment:
We're using uart0tx, uart0rx, uart1tx, and uart1rx as circular buffers, but because we haven't encapsulated the circular buffer logic anywhere it's repeated over and over, and it's not particularly clear (and I'm not certain it's fully understood) what interactions with these buffers will and will not cause race conditions. I've written a header/source file for a circular buffer that isn't fully documented to my usual level of rigor, but I think it'll be good enough for our purposes here. I also didn't actually try to compile this, and it's likely it's missing some includes. However, this is based on similar code I've written in the past so I'm pretty confident it should work. 
```
// RingBuffer.h

#ifndef __RING_BUFFER_H__
#define __RING_BUFFER_H__

// Forward definition of opaque ring buffer structure, defined in source file
// Other modules using this structure will not and do not need to know its contents 
typedef struct RingBuffer RingBuffer;

#define MAX_NUM_RING_BUFFERS 2

typedef enum RingBuffer__Status {
    RB__STATUS__SUCCESS = 0, /* Operation suceeded. */
    RB__STATUS__ERROR__NULL = -1, /* A required argument or a member of an argument was NULL */
    RB__STATUS__ERROR__FULL = -2, /* Did not write byte because buffer is full */
    RB__STATUS__ERROR__EMPTY = -3, /* Could not read byte because buffer is empty */
    RB__STATUS__ERROR__NOT_POWER_OF_TWO = -4, /* Buffer size was not a power of two */
    RB__STATUS__ERROR__ZERO_SIZE = -5, /* Buffer size was zero */
    RB__STATUS__ERROR__ALL_BUFFERS_USED = -6, /* All statically allocated RingBuffers already in use */

    RB__STATUS__ERROR__INTERNAL = -255 /* An unexpected internal error occurred. */
} RingBuffer__Status;

// Note that bufferSize MUST be a power of two
RingBuffer__Status RingBuffer__init(RingBuffer** rb, volatile uint8_t* buffer, size_t bufferSize);

int RingBuffer__full(const RingBuffer* rb);
int RingBuffer__empty(const RingBuffer* rb);

// put() will fail to write data if the buffer is full. get() will get the oldest byte
// from the buffer if the buffer is not empty.
//
// Either of these is safe to call in any context (in ISR or not) without needing to
// disable any interrupts.
//
// If put() is called in an interrupt and get() called outside of an interrupt, then
// these two functions can be used safely without disabling any interrupts. put() calls
// full(), which reads both the head and tail, and then may modify head. get() calls
// empty(), which reads both the head and tail, and then may modify tail. In this scenario
// calling put() is safe because it is called within the interrupt, so get() can't modify
// the head or tail while it is executed. Additionally, calling get() is safe because even
// though it *is* possible for head to be modified, this modification will only cause get()
// to not catch that a new element to be added to the buffer, but the next time get() is
// called it will be noticed.
//
// If get() is called in interrupt and put() is called outside an interrupt, then these
// functions are still safe to use. The worst thing that can happen is that during a 
// a call to put(), the get() is invoked from the interrupt. In this case, the get()
// call may free a slot in the buffer, but the call to put() won't see this free slot
// and the put() call will fail to write the byte to the buffer. Similarly, the get()
// call may not see a byte in the process of being written if the put() call is interrupted,
// but it is expected that after calling put() the code should make sure that whatever
// code calls get() is invoked (i.e. by enabling the appropriate interrupt).
RingBuffer__Status RingBuffer__put(RingBuffer* rb, uint8_t byte);
RingBuffer__Status RingBuffer__get(RingBuffer* rb, uint8_t* byte);

// putOverwrite() will overwrite the oldest byte in the buffer with the newest one if
// the buffer is full. getOverwrite() is the same as get(), but is implemented as a 
// separate function to make it clear why interrupts need to be disabled (see below).
//
// If using either of these in an interrupt, that same interrupt needs to be disabled
// before calling the other one. This is because putOverwrite() may modify both head
// *and* tail, so not disabling the interrupt will cause a race condition.
RingBuffer__Status RingBuffer__putOverwrite(RingBuffer* rb, uint8_t byte);
RingBuffer__Status RingBuffer__getOverwrite(RingBuffer* rb, uint8_t* byte);

#endif // #ifndef __RING_BUFFER_H__
```
```
// RingBuffer.c

#include "RingBuffer.h"

// Some definitions to identify magic numbers in the code
static const int TRUE = 1;
static const int FALSE = 0;

struct RingBuffer {
    volatile uint8_t* buffer;
    size_t bufferSize;
    volatile size_t head;
    volatile size_t tail;
}

static RingBuffer ALL_RING_BUFFERS[MAX_NUM_RING_BUFFERS];

RingBuffer__Status RingBuffer__init(RingBuffer** rb, volatile uint8_t* buffer, size_t bufferSize)
{
    static size_t rbIndex = 0;

    if (rb == NULL || buffer == NULL) {
        return RB__STATUS__ERROR__NULL;
    }

    if (bufferSize == 0U) {
        return RB__STATUS__ERROR__ZERO_SIZE;
    }

    // This condition is just a way of checking that bufferSize is a power of two
    if (((bufferSize - 1U) & bufferSize) != 0U) {
        return RB__STATUS__ERROR__NOT_POWER_OF_TWO;
    }

    if (rbIndex >= MAX_NUM_RING_BUFFERS) {
        return RB__STATUS__ERROR__ALL_BUFFERS_USED;
    }

    RingBuffer* thisRb = ALL_RING_BUFFERS + rbIndex;
    thisRb->buffer = buffer;
    thisRb->bufferSize = bufferSize;
    this->head = 0U;
    this->tail = 0U;
    *rb = thisRb;

    return RB__STATUS__SUCCESS;
}

int RingBuffer__full(const RingBuffer* rb)
{
    if (rb == NULL) {
        return FALSE;
    }

    return ((rb->head - rb->tail) == rb->bufferSize) ? TRUE : FALSE;
}

int RingBuffer__empty(const RingBuffer* rb)
{
    if (rb == NULL) {
        return FALSE;
    }

    return ((rb->head - rb->tail) == 0U) ? TRUE : FALSE;
}

RingBuffer__Status RingBuffer__put(RingBuffer* rb, uint8_t byte)
{
    if (rb == NULL || rb->buffer == NULL) {
        return RB__STATUS__ERROR__NULL;
    }

    if (RingBuffer__full(rb)) {
        return RB__STATUS__ERROR__FULL;
    }

    // Get the index of the head byte, wrapping if necessary
    // The operation below is equivalent to:
    //    rb->head % rb->bufferSize
    // and works because rb->bufferSize is guaranteed to be 
    // a power of two.
    size_t index = (rb->head & (rb->bufferSize - 1));
    
    rb->buffer[index] = byte;
    rb->head++;

    return RB__STATUS__SUCCESS;
}

RingBuffer__Status RingBuffer__get(RingBuffer* rb, uint8_t* byte)
{
    if (rb == NULL || rb->buffer == NULL) {
        return RB__STATUS__ERROR__NULL;
    }

    if (RingBuffer__empty(rb)) {
        return RB__STATUS__ERROR__EMPTY;
    }

    // Get the index of the head byte, wrapping if necessary
    // The operation below is equivalent to:
    //    rb->tail % rb->bufferSize
    // and works because rb->bufferSize is guaranteed to be 
    // a power of two.
    size_t index = (rb->tail & (rb->bufferSize - 1));
    
    *byte = rb->buffer[index];
    rb->tail++;

    return RB__STATUS__SUCCESS;
}

RingBuffer__Status RingBuffer__putOverwrite(RingBuffer* rb, uint8_t byte)
{
    if (rb == NULL || rb->buffer == NULL) {
        return RB__STATUS__ERROR__NULL;
    }

    int wasFull = RingBuffer__full(rb);

    // Get the index of the head byte, wrapping if necessary
    // The operation below is equivalent to:
    //    rb->head % rb->bufferSize
    // and works because rb->bufferSize is guaranteed to be 
    // a power of two.
    size_t index = (rb->head & (rb->bufferSize - 1));
    
    rb->buffer[index] = byte;
    rb->head++;

    // If the buffer was full before writing the byte, we need to
    // also increment the tail 
    if (wasFull) {
        rb->tail++;
    }

    return RB__STATUS__SUCCESS;
}

RingBuffer__Status RingBuffer__getOverwrite(RingBuffer* rb, uint8_t* byte)
{
    return RingBuffer__get(rb, byte);
}
```

Reply:
Sure. This seems like it would need to be part of a larger set of changes to simplify the interrupt handlers, too. I can probably work on this a bit this week.

Reply:
That's what I had in mind as well. I'm not sure I stressed it enough, but how much code is in the uart ISRs is perhaps the single biggest issue I've seen in this code, along with how the UART buffers are used in and out of the ISRs. That's a good thing, because the current implementation isn't wrong, per se. My belief is that moving the business logic out of the ISR and switching the buffers to use the code I posted here will make them more robust than they are now.